sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateProjection.scala

```diff
@@ -151,85 +151,96 @@ object GenerateProjection extends CodeGenerator[Seq[Expression], Projection] {
         s"""if (!nullBits[$i]) arr[$i] = c$i;"""
     }.mkString("\n      ")
 
+    val mutableStates = ctx.mutableStates.map {
+      case (jt, name, _) => s"private $jt $name;"
+    }.mkString("\n      ")
+
+    val initStates = ctx.mutableStates.zipWithIndex.map {
+      case ((jt, name, _), index) => s"$name = (${ctx.boxedType(jt)}) states[$index];"
+    }.mkString("\n        ")
+
     val code = s"""
-    public SpecificProjection generate($exprType[] expr) {
-      return new SpecificProjection(expr);
+    public SpecificProjection generate($exprType[] expr, Object[] states) {
+      return new SpecificProjection(expr, states);
     }
 
     class SpecificProjection extends ${classOf[BaseProject].getName} {
       private $exprType[] expressions = null;
+      $mutableStates
 
-      public SpecificProjection($exprType[] expr) {
+      public SpecificProjection($exprType[] expr, Object[] states) {
         expressions = expr;
+        $initStates
       }
 
       @Override
       public Object apply(Object r) {
-        return new SpecificRow(expressions, (InternalRow) r);
+        return new SpecificRow((InternalRow) r);
       }
-    }
 
-    final class SpecificRow extends ${classOf[MutableRow].getName} {
+      final class SpecificRow extends ${classOf[MutableRow].getName} {
```

Comment:
I made `SpecificRow` a inner class of `SpecificProjection` here, so that we can access these mutable states easily.